### DISCUSSION TITLE
mappedProperties inconsistency

### DISCUSSION THREAD

**Discussion by ghost** [JanusGraph/janusgraph#2586] [Q&A] [answered]
I need to retrieve all properties of edge/vertex

I am trying to `mappedProperties()` method by it doen't work consistent for me. Here is example
 
    mgmt = g.getGraph().openManagement()
    prop = mgmt.getPropertyKey('name');
    battled =  mgmt.getEdgeLabel('battled')
    mgmt.addProperties(battled, prop)
    battled.mappedProperties()
    mgmt.commit()

![image](https://user-images.githubusercontent.com/55307415/115716776-35493c00-a382-11eb-997f-a24980685218.png)

    mgmt = g.getGraph().openManagement()
    battled =  mgmt.getEdgeLabel('battled')
    battled.mappedProperties()
    mgmt.commit()

![image](https://user-images.githubusercontent.com/55307415/115716828-44c88500-a382-11eb-938f-5b67a4c9c709.png)

Am I doing something wrong?
Is there any other way to retrieve propertyKeys of edge/vertex from schema?

---------------------------------------
I am using _The Graph of the Gods_ in this example

**Comment by li-boxuan** [comment, marked as answer]
Did you enable "schema.constraints"?

**Reply by ghost** [reply]
It is disabled. I am using default configurations.
Does it required for this operation?

Thanks

**Reply by li-boxuan** [reply]
Yes. See https://docs.janusgraph.org/basics/schema/#schema-constraints

**Reply by ghost** [reply]
So when `schema.constraints` is disabled I am unable to get propertyKeys from schema, right?
Should I go through all data and retrieve properties with query like this, or is there a better way to do it?
```
g.V().
  hasLabel('god').
    properties().
    label().
    dedup().
    fold()
```

**Reply by li-boxuan** [reply]
Try

`mgmt.printPropertyKeys()`

or

`mgmt.printSchema()`

or

`mgmt.getRelationTypes(PropertyKey.class)`

**Reply by ghost** [reply]
Yes, I know about these methods. Them show all propertyKeys in schema, but I need only those that specific vertex/edge has. Is it possible?

**Reply by li-boxuan** [reply]
Not that I am aware of. You could run an OLAP job to collect the information you need.
I would suggest keeping your application level schema and keep that in sync with the real JanusGraph internal schema if possible.

**Reply by ghost** [reply]
Ok, thanks a lot. I will try to use OLAP
